### PR TITLE
Automatically verify that derive() lists are alphabetically ordered #…

### LINKED ISSUE #18172: Automatically verify that derive() lists are alphabetically ordered
Rust code that looks like `#[derive(Copy,Debug, ComputeSquaredDistance)]` should be reported as a style violation. The correct ordering is `#[derive(ComputeSquaredDistance, Copy, Debug)]`.

Code: `python/tidy/servo_tidy/tidy.py`
Tests: `python/tidy/servo_tidy_tests/rust_tidy.rs`, `python/tidy/servo_tidy_tests/test_tidy.rs`. Run the tests with `./mach test-tidy --self`.

### PR DESCRIPTION
**Pull request by davidcl**
Automatically verify that derive() lists are alphabetically ordered #18172

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #18172 (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18179)

### CODE REVIEW

File: python/tidy/servo_tidy/tidy.py [hunk 1]
```diff
@@ -493,6 +493,18 @@ def check_rust(file_name, lines):
             line = merged_lines + line
             merged_lines = ''
 
+        # derive traits should be alphabetically ordered
+        if line.startswith("#[derive"):
```
Comment (KiChjang) [inline python/tidy/servo_tidy/tidy.py]: Does `startswith` ignore initial whitespaces?

**Comment by jdm** [comment]
Wow, we have a lot of prior violations of this rule.

**Comment by davidcl** [comment]
Ouch sorry, it does not pass `./mach test-tidy`

File: python/tidy/servo_tidy/tidy.py (hunk 1 above)
Comment (davidcl) [inline python/tidy/servo_tidy/tidy.py]: no, I will fix that

**Comment by davidcl** [comment]
@jdm `test-tidy` does not report anything on my machine, what `./mach` command are you using ?
```
./mach test-tidy 
Checking the config file...
Checking directories for correct file extensions...
Checking files for tidiness...

tidy reported no errors.
```

**Comment by davidcl** [comment]
If the tidy derive is fully agreed by other contributors, I could also fix the existing codebase accordingly.

**Comment by jdm** [comment]
CC @nox

**Comment by nox** [comment]
Nice!

File: python/tidy/servo_tidy/tidy.py (hunk 1 above)
Comment (davidcl) [inline python/tidy/servo_tidy/tidy.py]: Done on https://github.com/servo/servo/pull/18179/commits/b33223952b0834d4a12c3e000aee1cdadf02f125

File: components/style/gecko/generated/structs_debug.rs [hunk 2]
```diff
@@ -10297,29 +10297,10 @@ pub mod root {
...
-        assert_eq!(::std::mem::size_of::<nsAutoString>() , 160usize , concat !
-                   ( "Size of: " , stringify ! ( nsAutoString ) ));
-        assert_eq! (::std::mem::align_of::<nsAutoString>() , 8usize , concat !
-                    ( "Alignment of " , stringify ! ( nsAutoString ) ));
-        assert_eq! (unsafe {
-                    & ( * ( 0 as * const nsAutoString ) ) . mStorage as *
-                    const _ as usize } , 32usize , concat ! (
-                    "Alignment of field: " , stringify ! ( nsAutoString ) ,
-                    "::" , stringify ! ( mStorage ) ));
+    pub type nsAutoStringN_self_type = u8;
```
Comment (jdm) [inline components/style/gecko/generated/structs_debug.rs]: Ooh, we should exclude this directory from these changes. This is generated code that's out of our control.

**Comment by KiChjang** [comment]
Also, merge commits are not accepted. Please rebase your PR against the master branch instead.

File: components/style/gecko/generated/structs_debug.rs (hunk 2 above)
Comment (davidcl) [inline components/style/gecko/generated/structs_debug.rs]: Done on e6a7943

**Comment by davidcl** [comment]
@KiChjang should I also prefer `rebase --squash` for 2f15ae2 e99c696 and 3a959b7 ? as I rewrote git history it should be cleaner to only have the tidy derive in one commit and rust code format on another.

**Comment by davidcl** [comment]
Review status: 0 of 196 files reviewed at latest revision, 2 unresolved discussions.

---

*[python/tidy/servo_tidy/tidy.py, line 497 at r1](https://reviewable.io:443/reviews/servo/servo/18179#-Ks5eSFCG6xAaeO8ij-x-r1-497:-KsCpqenWQDQPK87M4yA:b15q7ez) ([raw file](https://github.com/servo/servo/blob/61ce7dcb341667c1af4a88a1ae2ff098e1c11393/python/tidy/servo_tidy/tidy.py#L497)):*
<details><summary><i>Previously, davidcl (Clément DAVID) wrote…</i></summary><blockquote>

Done on https://github.com/servo/servo/pull/18179/commits/b33223952b0834d4a12c3e000aee1cdadf02f125
</blockquote></details>

Done. (reviewable.io did not parse the previous comment :) )

---

*[components/style/gecko/generated/structs_debug.rs, line 10300 at r6](https://reviewable.io:443/reviews/servo/servo/18179#-KsAccvouodGgUM0X-B8-r6-10300:-KsCp_G87qyAAMZY7Z9s:b15q7ez) ([raw file](https://github.com/servo/servo/blob/3138309ceee6029c9bac855d1a3d450e1239dcc4/components/style/gecko/generated/structs_debug.rs#L10300)):*
<details><summary><i>Previously, davidcl (Clément DAVID) wrote…</i></summary><blockquote>

Done on e6a7943
</blockquote></details>

Done. (reviewable.io did not parse the previous comment :) )

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/18179)*

**Comment by davidcl** [comment]
Review status: 0 of 196 files reviewed at latest revision, 2 unresolved discussions.

---

*[python/tidy/servo_tidy/tidy.py, line 497 at r1](https://reviewable.io:443/reviews/servo/servo/18179#-Ks5eSFCG6xAaeO8ij-x-r1-497:-KsCqJj34cHyufhgyKPW:b-896fix) ([raw file](https://github.com/servo/servo/blob/61ce7dcb341667c1af4a88a1ae2ff098e1c11393/python/tidy/servo_tidy/tidy.py#L497)):*
<details><summary><i>Previously, davidcl (Clément DAVID) wrote…</i></summary><blockquote>

Done. (reviewable.io did not parse the previous comment :) )
</blockquote></details>

Done.

---

*[components/style/gecko/generated/structs_debug.rs, line 10300 at r6](https://reviewable.io:443/reviews/servo/servo/18179#-KsAccvouodGgUM0X-B8-r6-10300:-KsCqIWCFv7WvJzdPAmL:b-896fix) ([raw file](https://github.com/servo/servo/blob/3138309ceee6029c9bac855d1a3d450e1239dcc4/components/style/gecko/generated/structs_debug.rs#L10300)):*
<details><summary><i>Previously, davidcl (Clément DAVID) wrote…</i></summary><blockquote>

Done. (reviewable.io did not parse the previous comment :) )
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/18179)*

**Comment by jdm** [comment]
There's a new tidy failure from rebasing, and this as well:
```
error[E0428]: a type named `RuleInclusion` has already been defined in this module

   --> /home/travis/build/servo/servo/components/style/stylist.rs:474:1

    |

81  | / pub enum RuleInclusion {

82  | |     /// Include rules for style sheets at all cascade levels.  This is the

83  | |     /// normal rule inclusion mode.

84  | |     All,

...   |

87  | |     DefaultOnly,

88  | | }

    | |_- previous definition of `RuleInclusion` here

...

474 | / pub enum RuleInclusion {

475 | |     /// Include rules for style sheets at all cascade levels.  This is the

476 | |     /// normal rule inclusion mode.

477 | |     All,

...   |

480 | |     DefaultOnly,

481 | | }

    | |_^ `RuleInclusion` already defined
```

**Comment by jdm** [comment]
Rather, travis reports a bunch of errors that look like incorrect rebasing.

File: components/style/stylist.rs [hunk 3]
```diff
@@ -76,9 +76,85 @@ struct DocumentCascadeData {
     precomputed_pseudo_element_decls: PerPseudoElementMap<Vec<ApplicableDeclarationBlock>>,
 }
 
-impl DocumentCascadeData {
```
Comment (jdm) [inline components/style/stylist.rs]: Here's the rebase issue.

**Review by jdm — APPROVED** [review]
This looks great! Let's squash the three commits you mentioned together, too.

**Comment by jdm** [comment]
```
./components/style/gecko/generated/pseudo_element_definition.rs:6: derivable traits list is not in alphabetical order

	expected: Clone, Debug, Eq, Hash, PartialEq

	found: Clone, Debug, PartialEq, Eq, Hash
```

File: components/style/stylist.rs (hunk 3 above)
Comment (davidcl) [inline components/style/stylist.rs]: Thanks, I re-run my upgrade script after the rebase (rather than before) and everything is fine now.

**Comment by jdm** [comment]
@bors-servo: r+
Thanks!

**Comment by jdm** [comment]
@bors-servo: retry
* git failure

**Pull request merged**